util/rowcodec/encoder.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rowcodec
+
+import (
+	"math"
+	"sort"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+// Encoder is used to encode a row.
+type Encoder struct {
+	row
+	tempColIDs []int64
+	values     []types.Datum
+	tempData   []byte
+	sc         *stmtctx.StatementContext
+}
+
+func (encoder *Encoder) reset() {
+	encoder.isLarge = false
+	encoder.numNotNullCols = 0
+	encoder.numNullCols = 0
+	encoder.data = encoder.data[:0]
+	encoder.tempColIDs = encoder.tempColIDs[:0]
+	encoder.values = encoder.values[:0]
+}
+
+func (encoder *Encoder) addColumn(colID int64, d types.Datum) {
+	if colID > 255 {
+		encoder.isLarge = true
+	}
+	if d.IsNull() {
+		encoder.numNullCols++
+	} else {
+		encoder.numNotNullCols++
+	}
+	encoder.tempColIDs = append(encoder.tempColIDs, colID)
+	encoder.values = append(encoder.values, d)
+}
+
+// Encode encodes a row from a datums slice.
+func (encoder *Encoder) Encode(colIDs []int64, values []types.Datum, buf []byte) ([]byte, error) {
+	encoder.reset()
+	for i, colID := range colIDs {
+		encoder.addColumn(colID, values[i])
+	}
+	return encoder.build(buf)
+}
+
+func (encoder *Encoder) build(buf []byte) ([]byte, error) {
+	r := &encoder.row
+	// Separate null and not-null column IDs.
+	numCols := len(encoder.tempColIDs)
+	nullIdx := numCols - int(r.numNullCols)
+	notNullIdx := 0
+	if r.isLarge {
+		encoder.initColIDs32()
+		encoder.initOffsets32()
+	} else {
+		encoder.initColIDs()
+		encoder.initOffsets()
+	}
+	for i, colID := range encoder.tempColIDs {
+		if encoder.values[i].IsNull() {
+			if r.isLarge {
+				r.colIDs32[nullIdx] = uint32(colID)
+			} else {
+				r.colIDs[nullIdx] = byte(colID)
+			}
+			nullIdx++
+		} else {
+			if r.isLarge {
+				r.colIDs32[notNullIdx] = uint32(colID)
+			} else {
+				r.colIDs[notNullIdx] = byte(colID)
+			}
+			encoder.values[notNullIdx] = encoder.values[i]
+			notNullIdx++
+		}
+	}
+	if r.isLarge {
+		largeNotNullSorter := (*largeNotNullSorter)(encoder)
+		sort.Sort(largeNotNullSorter)
```

Comment:
why do we need to sort `colIDs` and the `values` for each row?

Reply:
So we can do binary searches later.